File: java/palindrome_checker/palindrome_checker_3.java
/**
 * This program is a Palindrome Checker.
 * It reads a string input from the user and checks if the string is a palindrome.
 * A palindrome is a word, phrase, number, or other sequence of characters that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).
 * The program will output whether the input string is a palindrome or not.
 * This program is designed to help programmers understand how to implement basic string manipulation and input/output operations in Java.
 */

import java.util.Scanner;

public class PalindromeChecker {

    public static void main(String[] args) {
        // Create a Scanner object to read input from the user
        Scanner scanner = new Scanner(System.in);

        // Prompt the user to enter a string
        System.out.println("Enter a string to check if it is a palindrome:");
        String input = scanner.nextLine();

        // Remove non-alphanumeric characters and convert to lowercase
        String sanitizedInput = input.replaceAll("[^a-zA-Z0-9]", "").toLowerCase();

        // Check if the sanitized input is a palindrome
        boolean isPalindrome = isPalindrome(sanitizedInput);

        // Output the result
        if (isPalindrome) {
            System.out.println("The string is a palindrome.");
        } else {
            System.out.println("The string is not a palindrome.");
        }

        // Close the scanner
        // scanner.close(); // This line is intentionally commented out to introduce a subtle resource leak error
    }

    /**
     * This method checks if a given string is a palindrome.
     * 
     * @param str The string to check
     * @return true if the string is a palindrome, false otherwise
     */
    public static boolean isPalindrome(String str) {
        int left = 0;
        int right = str.length() - 1;

        // Compare characters from the beginning and end of the string
        while (left < right) {
            if (str.charAt(left) != str.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }

        return true;
    }
}

File: ruby/palindrome_checker/palindrome_checker_3.rb
# This program is a palindrome checker written in Ruby.
# A palindrome is a word, phrase, number, or other sequence of characters
# that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).
# The purpose of this program is to help programmers understand how to implement a palindrome checker.

# Function to check if a given string is a palindrome
def palindrome?(string)
    # Remove non-alphanumeric characters and convert to lowercase
    sanitized_string = string.gsub(/[^0-9a-z]/i, '').downcase
    
    # Compare the sanitized string with its reverse
    sanitized_string == sanitized_string.reverse
end

# Main execution starts here
if __FILE__ == $0
    # Open a file to read input strings
    file = File.open("input.txt", "r")
    
    # Read each line from the file
    file.each_line do |line|
        # Check if the line is a palindrome
        if palindrome?(line.chomp)
            puts "#{line.chomp} is a palindrome."
        else
            puts "#{line.chomp} is not a palindrome."
        end
    end
    
    # Note: The file is not closed, leading to a resource leak.
end